plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/DeploymentResourceMapper.java

```diff
@@ -387,341 +73,130 @@ public boolean isLoaded() {
-		for (IReplicationController rc : rcs) {
-			Map<String, String> deploymentSelector = rc.getReplicaSelector();
-			for (IService service : services) {
-				Map<String, String> serviceSelector = service.getSelector();
-				if (selectorsOverlap(serviceSelector, deploymentSelector)) {
-					createRelation(service, rc);
-				}
-			}
-		}
+	
+	private void load(String kind) {
+		List<IResource> resources = conn.getResources(kind, project.getName());
+		resources.forEach(r->add(r));
 	}
 
-	private void mapServicesToDeploymentConfigs(Collection<IService> services, Collection<IDeploymentConfig> dcs) {
-		for (IDeploymentConfig dc : dcs) {
-			Map<String, String> deploymentSelector = dc.getReplicaSelector();
-			for (IService service : services) {
-				Map<String, String> serviceSelector = service.getSelector();
-				if (selectorsOverlap(serviceSelector, deploymentSelector)) {
-					createRelation(service, dc);
-				}
-			}
-		}
-	}
 
-	private Collection<IBuild> getBuildsForPods(Collection<IPod> pods, Collection<IBuild> builds) {
-		Collection<IBuild> buildsForDeployment = new HashSet<IBuild>();
-		List<IDeploymentConfig> dcs = pods.stream().map(p -> getResourcesFor(p, ResourceKind.DEPLOYMENT_CONFIG))
-				.flatMap(l -> l.stream()).map(r -> (IDeploymentConfig) r).collect(Collectors.toList());
-		Map<String, Collection<IDeploymentConfig>> imageRefToDeployConfigs = mapImageRefToDeployConfigs(dcs);
-		for (IBuild build : builds) {
-			String buildImageRef = imageRef(build, this.project);
-			if (imageRefToDeployConfigs.containsKey(buildImageRef)) {
-				buildsForDeployment.add(build);
+	private void buildDeployments() {
+		if (state.compareAndSet(State.UNINITIALIZED, State.LOADING)) {
+			for (String kind : WatchManager.KINDS) {
+				load(kind);
 			}
-		}
-		return buildsForDeployment;
-	}
-
-	private Map<String, Collection<IDeploymentConfig>> mapImageRefToDeployConfigs(Collection<IDeploymentConfig> configs) {
-		return mapImageRefToDeployConfigs(new ConcurrentHashMap<>(), configs);
-	}
-	
-	private Map<String, Collection<IDeploymentConfig>> mapImageRefToDeployConfigs(Map<String, Collection<IDeploymentConfig>> map, Collection<IDeploymentConfig> configs){
-		for (IDeploymentConfig dc : configs) {
-			List<IDeploymentTrigger> imageChangeTriggers = filterImageChangeTriggers(dc);
-			for (IDeploymentTrigger trigger : imageChangeTriggers) {
-				String imageRef = imageRef((IDeploymentImageChangeTrigger) trigger, this.project);
-				if (!map.containsKey(imageRef)) {
-					map.put(imageRef, Collections.synchronizedSet(new HashSet<>()));
+			WatchManager.getInstance().stopWatch(project);
+			try {
+				ConnectionsRegistrySingleton.getInstance().removeListener(connectionListener);
+				Collection<IService> services = cache.getResourcesOf(ResourceKind.SERVICE);
+				Collection<Deployment> deployments = new ArrayList<>();
+				for (IService service : services) {
+					Deployment d = newDeployment(service);
+					deployments.add(d);
+					synchronized (cache) {
```

Comment:
I was trying to handle not needing to lock the entire object in order to do what needs to be done here.  Not sure if that is premature optimization or if it would be better just to lock the entire class.  Ideally, I would like to figure out a way to break this class up even further to promote more testing.  I think it would be accomplished via creating a series of static methods on a helper class.